lang/py/avro/io.py

```diff
@@ -180,15 +180,17 @@ def validate(expected_schema, datum):
             _DEBUG_VALIDATE_INDENT += 2
             if datum is not None and not datum:
                 print('{!s}<Empty>'.format(' ' * _DEBUG_VALIDATE_INDENT), file=sys.stderr)
-        result = _valid[expected_type](expected_schema, datum)
+        result = _valid[expected_type](expected_schema, datum, raise_on_error)
         if _DEBUG_VALIDATE:
             _DEBUG_VALIDATE_INDENT -= 2
             print('{!s}}} -> {!s}'.format(' ' * _DEBUG_VALIDATE_INDENT, result), file=sys.stderr)
     else:
-        result = _valid[expected_type](expected_schema, datum)
+        result = _valid[expected_type](expected_schema, datum, raise_on_error)
         if _DEBUG_VALIDATE:
             print('{!s}{!s}{!s}: {!s} -> {!s}'.format(' ' * _DEBUG_VALIDATE_INDENT,
                   expected_schema.type, name, type(datum).__name__, result), file=sys.stderr)
+        if raise_on_error and not result:
+            raise AvroTypeException(expected_schema, datum)
```

Comment:
Is there a difference if we just raise the exception from outside the function? i.e.
```
if not validate(schema, datum):
    raise AvroTypeException(schema, datum)
```

Reply:
Also does this work for unions?

Reply:
> Is there a difference if we just raise the exception from outside the function? i.e.
> 
> ```
> if not validate(schema, datum):
>     raise AvroTypeException(schema, datum)
> ```

Yes, because the schema validation is recursive. If we throw the exception outside of the validate function, it'll prettify the entire schema + datum. Inside the function, it'll throw the exception at the child node which invalidates the entire datum.

Reply:
> Also does this work for unions?

For array, map, union, record (will call these types 'collection' for now) I think they recursively call `validate` and will eventually land on a non-collection datum which will raise the Exception